### ISSUE TITLE
Restoring has_many associacions

### ISSUE THREAD

**Issue by tm-digitalpatrioten** [airblade/paper_trail#652]
Hi, 
I'm not sure if this is a bug, or if I'm simply doing something wrong. I have two very simple models:

```ruby
class Template < ActiveRecord::Base
  has_many :assets, dependent: :destroy
  has_paper_trail
  #also has a title
end

class Asset < ActiveRecord::Base
  belongs_to :template
  has_paper_trail
  has_attached_file :file, :path => ':class/:id/:filename.:extension'
  do_not_validate_attachment_file_type :file
end
```
This is the test I use at the moment:

``` ruby
    template = Template.create

    template.title = 'initial'
    template.assets.create()
    template.save

    title1 = template.title  #initial
    count1 = template.assets.count #1

    template.title = 'v2'
    template.assets.create()
    template.save

    title2 = template.title #v2
    count2 = template.assets.count #2

    template_old = template.versions.last.reify(:has_many => true)

    title3 = template_old.title #initial
    count3 = template_old.assets.count #2, should be 1

    template.title = 'v3'
    template.assets.last.destroy()
    template.save

    title4 = template.title #v3
    count4 = template.assets.count #1

    template_old = template.versions.last.reify(:has_many => true)

    title5 = template_old.title #v2
    count5 = template_old.assets.count #1, should be 2
```

Thanks in advance!

**Comment by tm-digitalpatrioten** [comment]
I used 
``` ruby
template.assets.create
```
to create the extra asset in version 2. The versions and the versions_associations table recorded the events for asset and template creation correctly (also with the fractals of seconds in the created timestamp).
In version_concern.rb in line 384 (reify_has_many_directly) the collections also contains two elements although it only should contain one. 


I also did another test where I made a 3rd version by destroying one of the assets and saving (asset destruction shows up in version table). 
So asset count in the v3 version is 1, where in the v2 version it is 2. The interesting thing in this case is, if I reify v2 and look at line 384, the collection contains 2 items but the model i receive after the reify is finished has only 1 item in its collection. I don't know if this is related. Just wanted to mention it.

**Comment by tm-digitalpatrioten** [comment]
This is the test I use at the moment, am I doing something wrong or expecting something wrong here?

``` ruby
    template = Template.create

    template.title = 'initial'
    template.assets.create()
    template.save

    title1 = template.title  #initial
    count1 = template.assets.count #1

    template.title = 'v2'
    template.assets.create()
    template.save

    title2 = template.title #v2
    count2 = template.assets.count #2

    template_old = template.versions.last.reify(:has_many => true)

    title3 = template_old.title #initial
    count3 = template_old.assets.count #2, should be 1

    template.title = 'v3'
    template.assets.last.destroy()
    template.save

    title4 = template.title #v3
    count4 = template.assets.count #1

    template_old = template.versions.last.reify(:has_many => true)

    title5 = template_old.title #v2
    count5 = template_old.assets.count #1, should be 2
```

**Comment by jaredbeck** [comment]
Instead of using `count`, try `length`.

```
template_old.assets.length
=> 1
template_old.assets.count
   (2.4ms)  SELECT COUNT(*) FROM "assets" WHERE "assets"."template_id" = ?  [["template_id", 1]]
=> 2
```

Note how `count` ignores the reified data and queries the database.

This is pretty surprising behavior, so I'm inclined to tentatively call it a bug, but I haven't the slightest idea how to fix it :cry:

Associations, and especially the reification of associations, is still an experimental feature in PaperTrail, and I don't recommend using it in production yet.

**Comment by tm-digitalpatrioten** [comment]
Hey Jared, 

thanks for answering. I just went through some more testing. After the reify, length and count behave in the same way (right after template_old is written).  I also tried eaching over the assets and it was identical with count/length.

I also had a look at the reify_has_many_directly function in version_concern.rb. In line 384 where the collection is put into the model, the collection also reflects the "live" state of the relation rather than the state of the relation at the time of the version. 
In 367 the live state of the association is loaded. The only point where elements (that have not existed yet) could be removed from the collection is 374. The strange thing here is that the versions hash in 362 was empty in all my tests. Therefore when mapping over the collection always the first case is selected in 372. I think there is probably something wrong here.

Hope this helped, let me know if i can help with some further testing or debugging.
